tree_sitter/binding/language.c

```diff
@@ -30,10 +33,119 @@ int language_init(Language *self, PyObject *args, PyObject *Py_UNUSED(kwargs)) {
+    void *ptr = NULL;
+    PyObject *ptr_method = NULL;
+    PyObject *ptr_result = NULL;
+    PyObject *cast_result = NULL;
+    PyObject *value_attr = NULL;
+
+    // Call .ptr() method on the managed pointer
+    ptr_method = PyObject_GetAttrString(managed_pointer, "ptr");
+    if (ptr_method == NULL) {
+        goto cleanup;
+    }
+    ptr_result = PyObject_CallObject(ptr_method, NULL);
+    if (ptr_result == NULL) {
+        goto cleanup;
+    }
+
+    // Call cast function
+    cast_result = PyObject_CallFunctionObjArgs(cast, ptr_result, c_void_p, NULL);
+    if (cast_result == NULL) {
+        goto cleanup;
+    }
+
+    // Get the 'value' attribute from the cast result
+    value_attr = PyObject_GetAttrString(cast_result, "value");
+    if (value_attr == NULL) {
+        goto cleanup;
+    }
+
+    // Convert the value attribute to a C void pointer
+    ptr = PyLong_AsVoidPtr(value_attr);
+
+cleanup:
+    Py_XDECREF(value_attr);
+    Py_XDECREF(cast_result);
+    Py_XDECREF(ptr_result);
+    Py_XDECREF(ptr_method);
+
+    if (PyErr_Occurred()) {
+        return NULL;
+    }
+
+    return ptr;
+}
+
+PyObject *language_from_wasm(PyTypeObject *cls, PyObject *args) {
+    ModuleState *state = (ModuleState *)PyType_GetModuleState(cls);
+    TSWasmError error;
+    TSWasmStore *wasm_store = NULL;
+    TSLanguage *language = NULL;
+    Language *self = NULL;
+    char *name;
+    PyObject *py_engine = NULL;
+    char *wasm;
+    Py_ssize_t wasm_length;
+    if (state->wasmtime_engine_type == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "wasmtime module is not loaded");
+        return NULL;
+    }
+    if (!PyArg_ParseTuple(args, "sO!y#:from_wasm", &name, state->wasmtime_engine_type, &py_engine, &wasm, &wasm_length)) {
+        return NULL;
+    }
+
+    TSWasmEngine *engine = (TSWasmEngine *)get_managed_pointer(state->ctypes_cast, state->c_void_p, py_engine);
+    if (engine == NULL) {
+        goto fail;
+    }
+    engine = wasmtime_engine_clone(engine);
+    if (engine == NULL) {
+        goto fail;
+    }
+
+    wasm_store = ts_wasm_store_new(engine, &error);
+    if (wasm_store == NULL) {
+        PyErr_Format(PyExc_RuntimeError, "Failed to create TSWasmStore: %s", error.message);
+        goto fail;
+    }
+
+    language = (TSLanguage *)ts_wasm_store_load_language(wasm_store, name, wasm, wasm_length, &error);
+    if (language == NULL) {
+        PyErr_Format(PyExc_RuntimeError, "Failed to load language: %s", error.message);
+        goto fail;
+    }
+
+    self = (Language *)cls->tp_alloc(cls, 0);
+    if (self == NULL) {
+        goto fail;
+    }
+
+    self->language = language;
+    self->wasm_engine = engine;
+    self->version = ts_language_version(self->language);
+#if HAS_LANGUAGE_NAMES
+    self->name = ts_language_name(self->language);
+#endif
+    return (PyObject *)self;
+
+fail:
+    if (engine != NULL) {
+        wasm_engine_delete(engine);
+    }
+    ts_language_delete(language);
```

Comment:
>_possible bug_

**✨ AI Suggestion:** In the `language_from_wasm` function, the `language` variable is not deleted in the error handling path if `cls->tp_alloc` fails. This can lead to a memory leak.
```suggestion
    self = (Language *)cls->tp_alloc(cls, 0);
    if (self == NULL) {
        ts_language_delete(language);
        goto fail;
    }

    self->language = language;
    self->wasm_engine = engine;
    self->version = ts_language_version(self->language);
#if HAS_LANGUAGE_NAMES
    self->name = ts_language_name(self->language);
#endif
    return (PyObject *)self;

fail:
    if (engine != NULL) {
        wasm_engine_delete(engine);
    }
```